presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java

```diff
@@ -306,40 +309,82 @@ public static int calculateLowWatermark(int maxPendingSplitsPerTask)
         for (RemoteTask task : existingTasks) {
             nodeToTaskMap.put(task.getNodeId(), task);
         }
-        List<ListenableFuture<?>> blockedFutures = blockedNodes.stream()
-                .map(Node::getNodeIdentifier)
-                .map(nodeToTaskMap::get)
-                .filter(Objects::nonNull)
-                .map(remoteTask -> remoteTask.whenSplitQueueHasSpace(spaceThreshold))
-                .collect(toImmutableList());
-        if (blockedFutures.isEmpty()) {
+
+        boolean blocked = true;
+        ImmutableList.Builder<FutureStateChange<?>.Listener> listeners = ImmutableList.builder();
+        for (Node blockedNode : blockedNodes) {
+            RemoteTask remoteTask = nodeToTaskMap.get(blockedNode.getNodeIdentifier());
+            if (remoteTask == null) {
+                continue;
+            }
+            Optional<FutureStateChange<?>.Listener> listener = remoteTask.whenSplitQueueHasSpace(spaceThreshold);
+            if (!listener.isPresent()) {
+                blocked = false;
+                break;
+            }
+            listeners.add(listener.get());
+        }
+        if (!blocked) {
             return immediateFuture(null);
         }
-        return getFirstCompleteAndCancelOthers(blockedFutures);
+        return getFirstCompleteAndCancelOthers(listeners.build());
     }
 
     public static ListenableFuture<?> toWhenHasSplitQueueSpaceFuture(List<RemoteTask> existingTasks, int spaceThreshold)
     {
         if (existingTasks.isEmpty()) {
             return immediateFuture(null);
         }
-        List<ListenableFuture<?>> stateChangeFutures = existingTasks.stream()
-                .map(remoteTask -> remoteTask.whenSplitQueueHasSpace(spaceThreshold))
-                .collect(toImmutableList());
-        return getFirstCompleteAndCancelOthers(stateChangeFutures);
+
+        boolean blocked = true;
+        ImmutableList.Builder<FutureStateChange<?>.Listener> listeners = ImmutableList.builder();
+        for (RemoteTask remoteTask : existingTasks) {
+            Optional<FutureStateChange<?>.Listener> listener = remoteTask.whenSplitQueueHasSpace(spaceThreshold);
+            if (!listener.isPresent()) {
+                blocked = false;
+                break;
+            }
+            listeners.add(listener.get());
+        }
+        if (!blocked) {
+            return immediateFuture(null);
+        }
+        return getFirstCompleteAndCancelOthers(listeners.build());
     }
 
-    private static ListenableFuture<?> getFirstCompleteAndCancelOthers(List<ListenableFuture<?>> blockedFutures)
+    /**
+     * Wait for any in the listeners to fire, and cancel all others to free up resources
+     */
+    private static ListenableFuture<?> getFirstCompleteAndCancelOthers(List<FutureStateChange<?>.Listener> listeners)
```

Comment:
This method is a bit complicated, it would be nice to have a test for this (if existing tests don't cover this already).